File: mobile/lib/core/services/api_client.dart
import 'package:dio/dio.dart';

import '../constants/app_constants.dart';
import 'secure_storage_service.dart';

class ApiClient {
  ApiClient({required SecureStorageService secureStorage, Dio? dio})
      : _secureStorage = secureStorage,
        dio = dio ?? Dio(BaseOptions(baseUrl: AppConstants.apiBaseUrl)) {
    this.dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (options, handler) async {
          options.headers['Accept'] = 'application/json';
          final token = await _secureStorage.readToken();
          if (token != null) {
            options.headers['Authorization'] = 'Bearer $token';
          }
          handler.next(options);
        },
        onError: (error, handler) async {
          if (error.response?.statusCode == 401) {
            await _secureStorage.clearToken();
          }
          handler.next(error);
        },
      ),
    );
  }

  final Dio dio;
  final SecureStorageService _secureStorage;
}

File: appReact/src/core/services/apiClient.ts
import axios, { AxiosInstance, InternalAxiosRequestConfig } from 'axios';
import { API_BASE_URL } from '../constants/appConstants';
import { SecureStorageService } from './secureStorageService';

export function createApiClient(secureStorage: SecureStorageService, http?: AxiosInstance): AxiosInstance {
  const client = http ?? axios.create({ baseURL: API_BASE_URL });

  client.interceptors.request.use(async (config: InternalAxiosRequestConfig) => {
    config.headers.set('Accept', 'application/json');
    const token = await secureStorage.readToken();
    if (token) {
      config.headers.set('Authorization', `Bearer ${token}`);
    }
    return config;
  });

  client.interceptors.response.use(
    (response) => response,
    async (error) => {
      if (error.response?.status === 401) {
        await secureStorage.clearToken();
      }
      return Promise.reject(error);
    }
  );

  return client;
}
